ipaserver/plugins/baseuser.py

```diff
@@ -694,3 +703,158 @@ class baseuser_remove_principal(LDAPRemoveAttribute):
     def pre_callback(self, ldap, dn, entry_attrs, attrs_list, *keys, **options):
         ensure_last_krbprincipalname(ldap, entry_attrs, *keys)
         return dn
+
+
+certmap_options = (
+    DNParam(
+        'issuer?',
+        cli_name='issuer',
+        label=_('Issuer'),
+        doc=_('Issuer of the certificate'),
+        flags=['virtual_attribute', 'no_create', 'no_update']
+    ),
+    DNParam(
+        'subject?',
+        cli_name='subject',
+        label=_('Subject'),
+        doc=_('Subject of the certificate'),
+        flags=['virtual_attribute', 'no_create', 'no_update']
+    ),
+    Bytes(
+        'usercertificate*', validate_certificate,
+        cli_name='certificate',
+        label=_('Certificate'),
+        doc=_('Base-64 encoded user certificate'),
+    ),
+)
+
+
+def _convert_to_x500(name):
+    """
+    Converts a (ipa) DN into a string representation following X500 order
+    """
+    if name:
+        return str(name.get_reverse())
+    return name
+
+
+def _build_mapdata(subject, issuer):
+    issuer = _convert_to_x500(issuer)
+    subject = _convert_to_x500(subject)
+    return u'X509:{issuer}{subject}'.format(
+        issuer='<I>{}'.format(issuer) if issuer else '',
+        subject='<S>{}'.format(subject) if subject else '',
+        )
+
+
+def _convert_options_to_certmap(options):
+    """
+    Converts options to ipacertmapdata
+
+    When --subject --issuer or --usercertificate options are used,
+    the value for ipacertmapdata is built from extracting subject and issuer,
+    converting their values to X500 ordering and using the format
+    X509:<I>issuer<S>subject
+    For instance:
+    X509:<I>O=DOMAIN,CN=Certificate Authority<S>O=DOMAIN,CN=user
+    A list of values can be returned if --usercertificate is used multiple
+    times, or in conjunction with --subject --issuer.
+    """
+    data = []
+    for item in options.get('ipacertmapdata', ()):
+        data.append(item)
+
+    if 'issuer' in options or 'subject' in options:
+        issuer = options.get('issuer')
+        subject = options.get('subject')
+        data.append(_build_mapdata(subject, issuer))
+
+    for dercert in options.get('usercertificate', ()):
+        cert = x509.load_certificate(dercert, x509.DER)
+        issuer = DN(cert.issuer)
+        subject = DN(cert.subject)
+        data.append(_build_mapdata(subject, issuer))
+
+    return data
+
+
+class baseuser_add_certmap(LDAPAddAttributeViaOption):
```

Comment:
This should be called `baseuser_add_certmapdata`, because the attribute is called `certmapdata` (actually it is called `data` now, but see my comment above). Ditto for `baseuser_remove_certmap`.